### DISCUSSION TITLE
New module not building - unable to import Oqtane.Shared/Client namespaces in razor pages

### DISCUSSION THREAD

**Discussion by lanthonyneville** [oqtane/oqtane.framework#1232] [Q&A]
This is my first shot at developing a module for Oqtane, and my first experience of Blazor too, so my issue may be due to ignorance about Blazor/Razor conventions :)

**The build errors**

In all my razor pages in my Client project, it is unable to resolve the usings to the namespaces in the Oqtane.Shared/Client NuGet references. It seems to think that @using Oqtane means @using <the project default namespace, which includes .Oqtane in the middle>. However, the references to the Oqtane.Shared and Client NuGet packages seem fine.

![image](https://user-images.githubusercontent.com/22200046/114079424-176bd980-98a2-11eb-965e-ed6f9214aaa3.png)

**Back-story:**

I took the Oqtane.Blogs solution as a starting point and began modifying it, retaining the basic project structure and NuGet references. All was going OK, until I tried installing the .nupkg and found that although the module installer reported that the package was installed (and DLLs copied, folder made under Modules etc), no new module definition had been created. Debugging Oqtane itself I found that the only DLLs inspected are ones with "oqtane" in the filename (see LoadModuleDefinitionsFromAssemblies() method). This is where things started to go wrong ....

I then changed both the assembly name and default namespace to the following format:
- Client project: assembly & default namespace = MyCompany.MyProduct.Oqtane.Client
- Server project: assembly & default namespace = MyCompany.MyProduct.Oqtane.Server
- Shared project: assembly & default namespace = MyCompany.MyProduct.Oqtane.Shared

I can't see how to resolve this. Do I have to name my assemblies & namespaces a particular way? I can't see what convention to use and I notice that the DNF.Projects solution has a convention somewhat like mine and builds OK.

**Comment by lanthonyneville** [comment]
So I finally got around the issue, and the Client project is building, but I don't get why what I did worked, and it has other side effects ...

I cleared the Default Namespace in the Client project. This resolved the issue in the _Imports.razor file where the imports of Oqtane namespaces seemed to be getting confused with namespaces in my Client project.

I then also changed the @namespace declaration in each of my .razor pages from MyCompany.MyProduct.Oqtane.Client to just Pages, and this resolved the imports of Oqtane namespaces there.

However the side effect of this is that, having now installed my module, the module definition is named "Pages" under Module Management in the Control Panel, not the name I have given it in the ModuleInfo class (in fact in the ModuleDefinitions table, most columns are blank):

![image](https://user-images.githubusercontent.com/22200046/114246853-9d178400-998b-11eb-8ca2-c42b0b3a9df5.png)

If someone can explain the rules for namespaces, and for setting the name of my module I would really appreciate it. Coming to Blazor and Oqtane as a newbie this is really confusing!

**Comment by leigh-pointer** [comment]
Hi @lanthonyneville 
It can seem daunting but it gets better. 
Have you tried using the Module creator, it scaffolds the module creation.  Using this to start module development would be my suggestion.

Also @sbwalker has created some great tutorials [https://www.youtube.com/watch?v=JPfUZPlRRCE](https://www.youtube.com/watch?v=JPfUZPlRRCE)

**Reply by lanthonyneville** [reply]
OK thanks. I just tried the Module Creator, and it worked, and gave me "sensible" namespaces. It seems to deliberately include ".Oqtane" in the assembly names, and sets the default namespace to the assembly name minus ".Oqtane".

**Reply by leigh-pointer** [reply]
@lanthonyneville great that you are up and running.
